R/factanal.R

```diff
@@ -0,0 +1,168 @@
+    cols_copy <- colnames(cleaned_df)
+    for (col in cols_copy) {
+      unique_val <- unique(cleaned_df[[col]])
+      if (length(unique_val) == 1) {
+        cleaned_df <- cleaned_df[colnames(cleaned_df) != col]
+      }
+    }
+    min_ncol <- 2
+    if (length(colnames(cleaned_df)) < min_ncol) {
+      if (length(grouped_cols) < 1) {
+        # If without group_by, throw error to display message.
+        stop("There are not enough columns after removing the columns with only NA or a single value.")
+      }
+      else {
+        # skip this group if less than 2 column is left. (We can't handle single column for now.)
+        return(NULL)
+      }
+    }
+    # "scale." is an argument name. There is no such operator like ".=". 
+    fit <- psych::fa(cleaned_df, nfactors = nfactors, fm = fm, scores = scores, rotate = rotate)
+    fit$correlation <- cor(cleaned_df) # For creating scree plot later.
+    fit$df <- filtered_df # add filtered df to model so that we can bind_col it for output. It needs to be the filtered one to match row number.
+    fit$grouped_cols <- grouped_cols
+    fit$sampled_nrow <- sampled_nrow
+    class(fit) <- c("fa_exploratory", class(fit))
+    fit
+  }
+
+  do_on_each_group(df, each_func, name = "model", with_unnest = FALSE)
+}
+
+# TODO: This is a code that worked with factanal. Need to update it for psych::fa to use if for Summary tab again.
+glance.fa_exploratory <- function(x, pretty.name = FALSE, ...) {
+  res <- broom:::glance.factanal(x) %>% dplyr::select(-n)
+  res <- res %>% dplyr::rename(`Factors`=n.factors, `Total Variance`=total.variance, `Chi-Square`=statistic, `P Value`=p.value, `Degree of Freedom`=df, `Method`=method, `Converged`=converged, `Number of Rows`=nobs)
+  res
+}
+
+#' extracts results from psych::fa object as a dataframe
+#' @export
+#' @param n_sample Sample number for biplot. Default 5000, which is the default of our scatter plot.
+#'        we use it for gathered_data for parallel coordinates too. sampling is applied before gather.
+tidy.fa_exploratory <- function(x, type="loadings", n_sample=NULL, pretty.name=FALSE, ...) {
+  if (type == "screeplot") {
+    eigen_res <- eigen(x$correlation, only.values = TRUE) # Cattell's scree plot is eigenvalues of correlation/covariance matrix.
+    res <- tibble::tibble(factor=1:length(eigen_res$values), eigenvalue=eigen_res$values)
+  }
+  else if (type == "variances") {
+    res <- tibble::as_tibble(t(x$Vaccounted)) %>% dplyr::mutate(Factor=as.factor(1:n()), `% Variance`=100*`Proportion Var`, `Cummulated % Variance`=100*`Cumulative Var`)
+  }
+  else if (type == "loadings") {
+    res <- broom:::tidy.factanal(x) # TODO: This just happens to work. Revisit.
+    res <- res %>% tidyr::pivot_longer(cols=c(starts_with("MR"), "uniqueness"), names_to="factor", values_to="value")
+    res <- res %>% dplyr::mutate(factor = case_when(factor=="uniqueness"~"Uniqueness", TRUE~stringr::str_replace(factor,"^MR","Factor ")))
```

Comment:
Add some sample of the column names starting with "MR" in a comment.

Reply:
Added comment with an example.